applications/StructuralMechanicsApplication/python_scripts/structural_response.py

```diff
@@ -220,7 +220,215 @@ def GetShapeGradient(self):
 
+    Attributes
+    ----------
+    primal_analysis : Primal analysis object of the response function
+    adjoint_analysis : Adjoint analysis object of the response function
+    response_function_utility: Cpp utilities object doing the actual computation of response value and gradient.
+    """
+    def __init__(self, identifier, project_parameters, model_part = None):
+        self.identifier = identifier
+
+        model = Model()
+        if model_part is not None:
+            model.AddModelPart(model_part)
+
+        # Create the primal solver
+        ProjectParametersPrimal = Parameters( open(project_parameters["primal_settings"].GetString(),'r').read() )
+        self.primal_analysis = structural_mechanics_analysis.StructuralMechanicsAnalysis(model, ProjectParametersPrimal)
+        self.primal_model_part_name = ProjectParametersPrimal["problem_data"]["model_part_name"].GetString()
+
+        # Create the adjoint solver
+        ProjectParametersAdjoint = Parameters( open(project_parameters["adjoint_settings"].GetString(),'r').read() )
+        ProjectParametersAdjoint["solver_settings"].AddValue("response_function_settings", project_parameters)
+
+        adjoint_model = Model()
+        # TODO find out why it is not possible to use the same model_part
+        self.adjoint_analysis = structural_mechanics_analysis.StructuralMechanicsAnalysis(adjoint_model, ProjectParametersAdjoint)
+        self.adjoint_model_part_name = ProjectParametersAdjoint["problem_data"]["model_part_name"].GetString()
+
+
+    def Initialize(self):
+        self.primal_analysis.Initialize()
+        self.adjoint_analysis.Initialize()
+
+        self.primal_model_part = self.primal_analysis.model.GetModelPart(self.primal_model_part_name)
+        self.adjoint_model_part = self.adjoint_analysis.model.GetModelPart(self.adjoint_model_part_name)
+
+        # TODO should be created here, not in solver!
+        self.response_function_utility = self.adjoint_analysis._GetSolver().response_function
+
+
+    def InitializeSolutionStep(self):
+        # synchronize the modelparts # TODO this should happen automatically
+        print("\n> Synchronize primal and adjoint modelpart for response:", self.identifier)
+        for node in self.adjoint_model_part.Nodes:
+            ref_node = self.primal_model_part.Nodes[node.Id]
+            node.X0 = ref_node.X0
+            node.Y0 = ref_node.Y0
+            node.Z0 = ref_node.Z0
+            node.X = ref_node.X
+            node.Y = ref_node.Y
+            node.Z = ref_node.Z
+
+
+        # Run the primal analysis.
+        # TODO if primal_analysis.status==solved: return
+        print("\n> Starting primal analysis for response:", self.identifier)
+        startTime = timer.time()
+        self.primal_analysis.time = self.primal_analysis._GetSolver().AdvanceInTime(self.primal_analysis.time)
+        self.primal_analysis.InitializeSolutionStep()
+        self.primal_analysis._GetSolver().Predict()
```

Comment:
Can we avoid calling these "private" functions? Can we call `self.primal_analysis.RunSolutionLoop()`?

Reply:
I want to do the advancing in time on my own. 
This is mainly because of the current design of the optimization application where these response functions are used. For each optimization iteration one static step is calculated.

With PR #2373 we will need an updated way of storing the results for each optimization iteration. Then it will become possible.

Reply:
You could achieve the same effect by incrementing `end_time` and calling `RunSolutionLoop`. With this you avoid calling private member functions from another class (i.e. `AnalysisStage`) and `primal_analysis` will maintain a meaningful value for `end_time` which aids the clear and correct use of the class (including internal logic).

Reply:
This is part of the requested changes of the current review.

Reply:
Done